pkg/lightning/backend/remote/chunk_sender.go

```diff
@@ -0,0 +1,398 @@
+			}
+			c.logger.Info("retry to put chunk", zap.Uint64("chunkID", nextChunkID))
+
+			data, err := sendRequest(ctx, c.httpClient, "PUT", url, buf)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			result = new(PutChunkResult)
+			err = json.Unmarshal(data, result)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if result.Canceled {
+				c.logger.Error("failed to put chunk, task is canceled", zap.String("error", result.Error))
+				return common.ErrLoadDataTaskCanceled.FastGenByArgs(c.e.loadDataTaskID, result.Error)
+			}
+
+			nextChunkID = result.HandledChunkID + 1
+		}
+	}
+
+	state := c.state.Load()
+	// We can clean the cache of chunks that the remote worker has flushed.
+	lastFlushedChunkID := state.FlushedChunkID + 1
+	for lastFlushedChunkID <= result.FlushedChunkID {
+		err := c.chunksCache.clean(lastFlushedChunkID)
+		if err != nil {
+			c.logger.Info("failed to clean chunk cache",
+				zap.Uint64("chunkID", lastFlushedChunkID),
+				zap.Error(err))
+		}
+		lastFlushedChunkID++
+	}
+
+	c.state.Store(&chunksState{
+		FlushedChunkID: result.FlushedChunkID,
+		HandledChunkID: result.HandledChunkID,
+	})
+	return nil
+}
+
+// sendFlushToRemote send flush requst to remote worker.
+//
+// The remote worker should return `FlushedChunkID`, which should be equal to `LastChunkID`.
+// If not, it means that the remote worker has restarted and we should resend the chunk from
+// `FlushedChunkID` to the `LastChunkID`.
+func (c *chunkSender) sendFlushToRemote(ctx context.Context) error {
+	for {
+		url := fmt.Sprintf(flushURL, c.e.addr, c.e.clusterID, c.e.loadDataTaskID, c.id)
+
+		data, err := sendRequest(ctx, c.httpClient, "POST", url, nil)
+		if err != nil {
+			c.logger.Error("failed to flush", zap.Error(err))
+			return errors.Trace(err)
+		}
+
+		result := new(FlushResult)
+		err = json.Unmarshal(data, result)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		if result.Canceled || result.Error != "" {
+			c.logger.Error("failed to flush",
+				zap.Bool("canceled", result.Canceled),
+				zap.String("error", result.Error))
+			return common.ErrLoadDataTaskCanceled.FastGenByArgs(c.e.loadDataTaskID, result.Error)
+		}
+
+		if result.Finished {
+			c.logger.Info("load data task finished")
+			return nil
+		}
+
+		// Make sure all chunks are flushed in remote worker.
+		if result.FlushedChunkID == c.getLastChunkID() {
+			// Clean the cache of chunks that have been flushed and update the state.
+			state := c.state.Load()
+			lastFlushedChunkID := state.FlushedChunkID + 1
+			for lastFlushedChunkID <= result.FlushedChunkID {
+				err := c.chunksCache.clean(lastFlushedChunkID)
+				if err != nil {
+					c.logger.Warn("failed to clean chunk cache", zap.Uint64("chunkID", lastFlushedChunkID), zap.Error(err))
+				}
+				lastFlushedChunkID++
+			}
+			state.FlushedChunkID = result.FlushedChunkID
+
+			c.state.Store(state)
+			return nil
+		}
+
+		// The remote worker missed some chunks, we need to retry putting chunks.
```

Comment:
if we go here, how will we break the loop started at line 337? I think the only happy exit is line 364, but we didn't send actual new chunks with data in this loop, so remote server will not advance "flushed chunk ID" and line 364 will never be true

Reply:
The loop will break when the `result.FlushedChunkID == c.getLastChunkID`.

Reply:
Oh I see this is a tricky usage, the chunk will be sent again in `putEmptyChunk`

Reply:
yesh